File: src/go/pkg/netipc/service/raw/server_unix.go
//go:build unix

package raw

import (
	"sync"
	"sync/atomic"

	"github.com/netdata/netdata/go/plugins/pkg/netipc/protocol"
	"github.com/netdata/netdata/go/plugins/pkg/netipc/transport/posix"
)

// Server is an internal managed server bound to one expected request kind.
// Supports multiple concurrent client sessions up to workerCount.
type Server struct {
	runDir                       string
	serviceName                  string
	config                       posix.ServerConfig
	expectedMethodCode           uint16
	handler                      DispatchHandler
	running                      atomic.Bool
	learnedRequestPayloadBytes   atomic.Uint32
	learnedResponsePayloadBytes  atomic.Uint32
	requestPayloadGrowthCeiling  uint32
	responsePayloadGrowthCeiling uint32
	nextSessionID                atomic.Uint64
	workerCount                  int
	wg                           sync.WaitGroup
	listenerMu                   sync.Mutex
	listener                     *posix.Listener
}

// NewServer creates a new managed server. workerCount limits the
// maximum number of concurrent client sessions (default 1 if <= 0).
func NewServer(
	runDir, serviceName string,
	config posix.ServerConfig,
	expectedMethodCode uint16,
	handler DispatchHandler,
) *Server {
	return NewServerWithWorkers(runDir, serviceName, config, expectedMethodCode, handler, defaultServerWorkerCount)
}

// NewServerWithWorkers creates a server with an explicit worker count limit.
func NewServerWithWorkers(
	runDir, serviceName string,
	config posix.ServerConfig,
	expectedMethodCode uint16,
	handler DispatchHandler,
	workerCount int,
) *Server {
	s := &Server{config: config}
	s.initCommon(
		runDir,
		serviceName,
		expectedMethodCode,
		handler,
		workerCount,
		config.MaxRequestPayloadBytes,
		config.MaxResponsePayloadBytes,
	)
	return s
}

const posixShmProfiles = protocol.ProfileSHMHybrid | protocol.ProfileSHMFutex

func (s *Server) prepareAcceptConfig() (uint64, posix.ServerConfig, *posix.ShmContext, bool) {
	sessionID := s.nextSessionID.Add(1)
	cfg := s.config
	cfg.MaxRequestPayloadBytes = s.learnedRequestPayloadBytes.Load()
	cfg.MaxResponsePayloadBytes = s.learnedResponsePayloadBytes.Load()

	if cfg.SupportedProfiles&posixShmProfiles == 0 {
		return sessionID, cfg, nil, true
	}

	shm, err := posix.ShmServerCreate(
		s.runDir, s.serviceName, sessionID,
		cfg.MaxRequestPayloadBytes+uint32(protocol.HeaderSize),
		cfg.MaxResponsePayloadBytes+uint32(protocol.HeaderSize),
	)
	if err == nil {
		return sessionID, cfg, shm, true
	}

	cfg.SupportedProfiles &^= posixShmProfiles
	cfg.PreferredProfiles &^= posixShmProfiles
	if cfg.SupportedProfiles == 0 {
		return sessionID, cfg, nil, false
	}

	return sessionID, cfg, nil, true
}

// Run starts the acceptor loop. Blocking. Accepts clients, spawns a
// goroutine per session (up to workerCount concurrently).
// Returns when Stop() is called or on fatal error.
func (s *Server) Run() error {
	posix.ShmCleanupStale(s.runDir, s.serviceName)

	listener, err := posix.Listen(s.runDir, s.serviceName, s.config)
	if err != nil {
		return err
	}
	s.setListener(listener)
	defer s.closeListener(listener)

	s.running.Store(true)

	/* Semaphore channel limits concurrent sessions */
	sem := make(chan struct{}, s.workerCount)

	for s.running.Load() {
		// Poll the listener fd before blocking on accept
		ready := pollFd(listener.Fd(), serverPollTimeoutMs)
		if ready < 0 {
			break
		}
		if ready == 0 {
			continue
		}

		sessionID, acceptCfg, precreatedShm, ok := s.prepareAcceptConfig()
		if !ok {
			s.retryAcceptAfter(nil)
			continue
		}

		session, err := listener.AcceptWithConfig(sessionID, acceptCfg)
		if err != nil {
			if !s.retryAcceptAfter(func() {
				if precreatedShm != nil {
					precreatedShm.ShmDestroy()
				}
			}) {
				break
			}
			continue
		}

		if !s.acquireWorkerSlot(sem, func() {
			if precreatedShm != nil {
				precreatedShm.ShmDestroy()
			}
		}, session.Close) {
			continue
		}

		var shm *posix.ShmContext
		if session.SelectedProfile == protocol.ProfileSHMHybrid ||
			session.SelectedProfile == protocol.ProfileSHMFutex {
			if precreatedShm == nil {
				session.Close()
				<-sem
				continue
			}
			shm = precreatedShm
		} else if precreatedShm != nil {
			precreatedShm.ShmDestroy()
		}

		s.startSessionWorker(sem, func() {
			s.handleSession(session, shm)
		})
	}

	// Wait for all active session goroutines to finish
	s.wg.Wait()

	return nil
}

// Stop signals the server to stop and unblocks Accept by closing the listener.
func (s *Server) Stop() {
	s.running.Store(false)
	s.closeListener(nil)
}

func (s *Server) setListener(listener *posix.Listener) {
	s.listenerMu.Lock()
	s.listener = listener
	s.listenerMu.Unlock()
}

func (s *Server) closeListener(listener *posix.Listener) {
	s.listenerMu.Lock()
	target := listener
	if target == nil {
		target = s.listener
	}
	if target != nil && s.listener == target {
		s.listener = nil
	}
	s.listenerMu.Unlock()

	if target != nil {
		target.Close()
	}
}

func (s *Server) handleSession(session *posix.Session, shm *posix.ShmContext) {
	s.handleServerSession(serverSessionOps{
		maxRequestPayloadBytes:  session.MaxRequestPayloadBytes,
		maxResponsePayloadBytes: session.MaxResponsePayloadBytes,
		receive: func(recvBuf []byte) (protocol.Header, []byte, serverReceiveAction) {
			if shm != nil {
				mlen, err := shm.ShmReceive(recvBuf, serverPollTimeoutMs)
				if err != nil {
					if err == posix.ErrShmTimeout {
						return protocol.Header{}, nil, serverReceiveContinue
					}
					return protocol.Header{}, nil, serverReceiveStop
				}
				if mlen < protocol.HeaderSize {
					return protocol.Header{}, nil, serverReceiveStop
				}
				hdr, err := protocol.DecodeHeader(recvBuf[:mlen])
				if err != nil {
					return protocol.Header{}, nil, serverReceiveStop
				}
				return hdr, recvBuf[protocol.HeaderSize:mlen], serverReceiveOK
			}

			ready := pollFd(session.Fd(), serverPollTimeoutMs)
			if ready < 0 {
				return protocol.Header{}, nil, serverReceiveStop
			}
			if ready == 0 {
				return protocol.Header{}, nil, serverReceiveContinue
			}

			hdr, payload, err := session.Receive(recvBuf)
			if err != nil {
				return protocol.Header{}, nil, serverReceiveStop
			}
			return hdr, payload, serverReceiveOK
		},
		send: func(respHdr *protocol.Header, payload []byte, msgBuf *[]byte) error {
			if shm == nil {
				return session.Send(respHdr, payload)
			}
			msg, err := serverEncodeSharedResponse(respHdr, payload, msgBuf)
			if err != nil {
				return err
			}
			return shm.ShmSend(msg)
		},
		close: func() {
			if shm != nil {
				shm.ShmDestroy()
			}
			session.Close()
		},
	})
}

File: src/crates/netipc/src/service/raw/server_unix.rs
use super::common::SERVER_POLL_TIMEOUT_MS;
use super::server::{ManagedServer, ServerConfig};
use super::server_session_unix::{handle_session_threaded, poll_fd};
use crate::protocol::NipcError;
#[cfg(target_os = "linux")]
use crate::protocol::{HEADER_SIZE, PROFILE_SHM_FUTEX, PROFILE_SHM_HYBRID};
use crate::transport::posix::UdsListener;
#[cfg(target_os = "linux")]
use crate::transport::shm::ShmContext;
use std::sync::atomic::Ordering;

impl ManagedServer {
    /// Run the acceptor loop. Blocking. Accepts clients, spawns a
    /// thread per session (up to worker_count concurrent sessions).
    ///
    /// Returns when `stop()` is called or on fatal error.
    #[cfg(unix)]
    pub fn run(&mut self) -> Result<(), NipcError> {
        #[cfg(target_os = "linux")]
        crate::transport::shm::cleanup_stale(&self.run_dir, &self.service_name);

        let listener = UdsListener::bind(
            &self.run_dir,
            &self.service_name,
            self.server_config.clone(),
        )
        .map_err(|_| NipcError::BadLayout)?;

        self.running.store(true, Ordering::Release);

        let mut session_threads: Vec<std::thread::JoinHandle<()>> = Vec::new();

        while self.running.load(Ordering::Acquire) {
            let ready = poll_fd(listener.fd(), SERVER_POLL_TIMEOUT_MS as i32);
            if ready < 0 {
                break;
            }
            if ready == 0 {
                session_threads.retain(|t| !t.is_finished());
                continue;
            }

            let (session_id, accept_cfg, precreated_shm, ready) = self.prepare_unix_accept();
            if !ready {
                std::thread::sleep(std::time::Duration::from_millis(10));
                continue;
            }

            let session = match listener.accept_with_config(session_id, accept_cfg) {
                Ok(s) => s,
                Err(_) => {
                    #[cfg(target_os = "linux")]
                    if let Some(mut shm) = precreated_shm {
                        shm.destroy();
                    }
                    if !self.running.load(Ordering::Acquire) {
                        break;
                    }
                    std::thread::sleep(std::time::Duration::from_millis(10));
                    continue;
                }
            };

            session_threads.retain(|t| !t.is_finished());
            if session_threads.len() >= self.worker_count {
                #[cfg(target_os = "linux")]
                if let Some(mut shm) = precreated_shm {
                    shm.destroy();
                }
                drop(session);
                continue;
            }

            #[cfg(target_os = "linux")]
            let shm = match self.finalize_unix_shm(&session, precreated_shm) {
                Some(shm) => Some(shm),
                None if session.selected_profile == PROFILE_SHM_HYBRID
                    || session.selected_profile == PROFILE_SHM_FUTEX =>
                {
                    drop(session);
                    continue;
                }
                None => None,
            };
            #[cfg(not(target_os = "linux"))]
            let shm: Option<()> = None;

            let expected_method_code = self.expected_method_code;
            let handler = self.handler.clone();
            let running = self.running.clone();
            let learned_request_payload_bytes = self.learned_request_payload_bytes.clone();
            let learned_response_payload_bytes = self.learned_response_payload_bytes.clone();
            let request_payload_growth_ceiling = self.request_payload_growth_ceiling;
            let response_payload_growth_ceiling = self.response_payload_growth_ceiling;

            let t = std::thread::spawn(move || {
                handle_session_threaded(
                    session,
                    #[cfg(target_os = "linux")]
                    shm,
                    #[cfg(not(target_os = "linux"))]
                    shm,
                    expected_method_code,
                    handler,
                    running,
                    learned_request_payload_bytes,
                    learned_response_payload_bytes,
                    request_payload_growth_ceiling,
                    response_payload_growth_ceiling,
                );
            });
            session_threads.push(t);
        }

        for t in session_threads {
            let _ = t.join();
        }

        Ok(())
    }

    #[cfg(target_os = "linux")]
    fn prepare_unix_accept(&mut self) -> (u64, ServerConfig, Option<ShmContext>, bool) {
        let session_id = self.next_session_id;
        self.next_session_id += 1;

        let mut cfg = self.server_config.clone();
        cfg.max_request_payload_bytes = self.learned_request_payload_bytes.load(Ordering::Acquire);
        cfg.max_response_payload_bytes =
            self.learned_response_payload_bytes.load(Ordering::Acquire);

        let shm_profiles = cfg.supported_profiles & (PROFILE_SHM_HYBRID | PROFILE_SHM_FUTEX);
        if shm_profiles == 0 {
            return (session_id, cfg, None, true);
        }

        match ShmContext::server_create(
            &self.run_dir,
            &self.service_name,
            session_id,
            cfg.max_request_payload_bytes + HEADER_SIZE as u32,
            cfg.max_response_payload_bytes + HEADER_SIZE as u32,
        ) {
            Ok(ctx) => (session_id, cfg, Some(ctx), true),
            Err(_) => {
                cfg.supported_profiles &= !(PROFILE_SHM_HYBRID | PROFILE_SHM_FUTEX);
                cfg.preferred_profiles &= !(PROFILE_SHM_HYBRID | PROFILE_SHM_FUTEX);
                (session_id, cfg.clone(), None, cfg.supported_profiles != 0)
            }
        }
    }

    #[cfg(target_os = "linux")]
    fn finalize_unix_shm(
        &self,
        session: &crate::transport::posix::UdsSession,
        mut shm: Option<ShmContext>,
    ) -> Option<ShmContext> {
        let profile = session.selected_profile;
        if profile != PROFILE_SHM_HYBRID && profile != PROFILE_SHM_FUTEX {
            if let Some(ref mut ctx) = shm {
                ctx.destroy();
            }
            return None;
        }
        shm
    }
}
